### ISSUE TITLE
Analyzing diploid mESC data from Nagano et al. (2017)

### ISSUE THREAD

**Issue by tarak77** [lh3/hickit#5]
Hello again,

In Dip-C paper, there is a cross section view of the diploid interphase mESC cell 1CDX1-413 and some other cells. The model and its analysis looks good. 
In the paper by Nagano et al. (2017) an intercross of Mus musculus 129/SV-Jae and Mus musculus castaneus cells were used. I had a big favor to ask- will it be possible to provide the source for the phased SNPs in VCF file for the mESC diploid cells? and an example pipeline used to generate the 3D models of these diploid mouse cells ? will definitely be  great.

I really appreciate your help!

**Comment by lh3** [comment]
You have to ask @tanlongzhi on this one. I haven't run hickit on these cells.

**Comment by tarak77** [comment]
Okay, thanks

**Comment by tanlongzhi** [comment]
I have provided instructions for generating the SNP file for our diploid mESC re-analysis in the [corresponding thread](https://github.com/tanlongzhi/dip-c/issues/2#issuecomment-418700031) in [dip-c](https://github.com/tanlongzhi/dip-c).

I'll keep this thread open for now in case you encounter problems running this repo on diploid mESCs.

**Comment by tarak77** [comment]
Hi @lh3 @tanlongzhi ,
Coming back to working on mouse diploid cell data from FASTQ files from the same paper by Nagano et al.(2017), I have the FASTQ files for mESC cell 1CDX1-25 in this format:
```
@HWI-1KL136:238:C39NTACXX:3:1101:1621:1993 4:N:0:TAAGGCGA_TATCCTCT
GTTCAGTACAGTGCAGAATGGAAGGAAGGAGACAGCCACAATGTCATGCTG
+
?@BDDD?DHFHFFE?<FDIGGCD@GGJII>CEFHGIBAD@D@GDDGGGIGH
@HWI-1KL136:238:C39NTACXX:3:1101:2162:1982 4:N:0:TAAGGCGA_TATCCTCT
GTTCATATTGTTGTTCCTCCTATAGGGTTGCAAACCCCTTCAGCTCCTTGG
+
@BBFFFFFHAFHHJJIJIIIIEHIIIJGHIIJBFFGHIJJJIJJJJJGDGB
@HWI-1KL136:238:C39NTACXX:3:1101:3680:1994 4:N:0:TAAGGCGA_TATCCTCT
GTCAACATGCAGTGGTAGAAAATAATTCTCCCCTTTTCAGAGCAAGAAACT
+
```

and for the reference genome I use the `http://hgdownload.cse.ucsc.edu/goldenPath/mm9/chromosomes/` to get `chr*.fa` files

Then, I implement the first step in hickit,and get 
```
wg-dhcp174d160d214:hickit tarakshisode$ bwa mem  ../nuc_processing-master/chromosomes/*.fa ../nuc_processing-master/data/lane4559_AGGCAGAA_CTCTCTAT_1CDX1-25_L001_R1.fq.gz ../nuc_processing-master/data/lane4559_AGGCAGAA_CTCTCTAT_1CDX1-25_L001_R4.fq.gz | gzip > ./1CDX1-25/aln.sam.gz

Usage: bwa mem [options] <idxbase> <in1.fq> [in2.fq]

Algorithm options:

       -t INT        number of threads [1]
       -k INT        minimum seed length [19]
       -w INT        band width for banded alignment [100]
       -d INT        off-diagonal X-dropoff [100]
       -r FLOAT      look for internal seeds inside a seed longer than {-k} * FLOAT [1.5]
       -y INT        seed occurrence for the 3rd round seeding [20]
       -c INT        skip seeds with more than INT occurrences [500]
       -D FLOAT      drop chains shorter than FLOAT fraction of the longest overlapping chain [0.50]
       -W INT        discard a chain if seeded bases shorter than INT [0]
       -m INT        perform at most INT rounds of mate rescues for each read [50]
       -S            skip mate rescue
       -P            skip pairing; mate rescue performed unless -S also in use

Scoring options:

       -A INT        score for a sequence match, which scales options -TdBOELU unless overridden [1]
       -B INT        penalty for a mismatch [4]
       -O INT[,INT]  gap open penalties for deletions and insertions [6,6]
       -E INT[,INT]  gap extension penalty; a gap of size k cost '{-O} + {-E}*k' [1,1]
       -L INT[,INT]  penalty for 5'- and 3'-end clipping [5,5]
       -U INT        penalty for an unpaired read pair [17]

       -x STR        read type. Setting -x changes multiple parameters unless overridden [null]
                     pacbio: -k17 -W40 -r10 -A1 -B1 -O1 -E1 -L0  (PacBio reads to ref)
                     ont2d: -k14 -W20 -r10 -A1 -B1 -O1 -E1 -L0  (Oxford Nanopore 2D-reads to ref)
                     intractg: -B9 -O16 -L5  (intra-species contigs to ref)

Input/output options:

       -p            smart pairing (ignoring in2.fq)
       -R STR        read group header line such as '@RG\tID:foo\tSM:bar' [null]
       -H STR/FILE   insert STR to header if it starts with @; or insert lines in FILE [null]
       -o FILE       sam file to output results to [stdout]
       -j            treat ALT contigs as part of the primary assembly (i.e. ignore <idxbase>.alt file)
       -5            for split alignment, take the alignment with the smallest coordinate as primary
       -q            don't modify mapQ of supplementary alignments
       -K INT        process INT input bases in each batch regardless of nThreads (for reproducibility) []

       -v INT        verbosity level: 1=error, 2=warning, 3=message, 4+=debugging [3]
       -T INT        minimum score to output [30]
       -h INT[,INT]  if there are <INT hits with score >80% of the max score, output all in XA [5,200]
       -a            output all alignments for SE or unpaired PE
       -C            append FASTA/FASTQ comment to SAM output
       -V            output the reference FASTA header in the XR tag
       -Y            use soft clipping for supplementary alignments
       -M            mark shorter split hits as secondary

       -I FLOAT[,FLOAT[,INT[,INT]]]
                     specify the mean, standard deviation (10% of the mean if absent), max
                     (4 sigma from the mean if absent) and min of the insert size distribution.
                     FR orientation only. [inferred]

Note: Please read the man page for detailed description of the command line and options.
```

Am I doing it right? Any help will be good.

P.S. I have prebuilt mm9 genome indices which I could use in pipeline? If yes, will it possible to describe the steps?

**Comment by tanlongzhi** [comment]
You cannot use bowtie2's index for bwa. You must build a bwa index for your genome's `.fa` file (by the way, I don't think you can use multiple `.fa` files like `chr*.fa`).

The mouse genome I used was downloaded from GENCODE, whose current `.fa` file can be downloaded from this FTP link:
```
ftp://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_mouse/release_M18/GRCm38.p6.genome.fa.gz
```

After unzipping this file, you will need to build a bwa index:
```sh
bwa index -a bwtsw genome.fa
```

**Comment by tarak77** [comment]
It generate five files (.amb, .ann, .bwt, .pac, .sa files), what to do after that?

**Comment by tanlongzhi** [comment]
You can now proceed to map your reads:
```sh
bwa mem -p genome.fa read1.fq.gz read2.fq.gz | gzip > aln.sam.gz
```

**Comment by tarak77** [comment]
Okay, its mapping. But what was the use of the produced five files?

**Comment by tanlongzhi** [comment]
Great.

The five files (the bwa index for the `.fa` file) are essential for running bwa to map reads. Please refer to the [bwa README](https://github.com/lh3/bwa#introduction) for details on the index.

**Comment by tarak77** [comment]
Got it!
When I try to find the contact pairs using the phased_SNP.tsv file, it gives the following error:

```
(py2) wg-dhcp174d160d214:hickit tarakshisode$ ./hickit.js sam2seg -v ./cast_129.phased_SNP.tsv ./1CDX1-25/aln.sam.gz | ./hickit.js chronly - | gzip > ./1CDX1-25/contacts.seg.gz
env: env: k8: No such file or directory
k8: No such file or directory
```

**Comment by tanlongzhi** [comment]
You need to install a Javascript shell to run the script `hickit.js`. We use [k8](https://github.com/attractivechaos/k8); but you can choose another shell if you prefer.

**Comment by tarak77** [comment]
I am kinda new to javascript, after downloading the precompiled binaries for k8, could you help with how to run the command?

**Comment by tanlongzhi** [comment]
Suppose you're using Linux, then you can run `hickit.js` by:
```sh
<path_to_k8_download>/k8-Linux hickit.js
```

If you copy `k8-Linux` to your own path (for example `~/bin` on Harvard Research Computing) and rename as `k8`, you can also run `hickit.js` directly by:
```
hickit.js

**Comment by tarak77** [comment]
Ok. I am running node.js on my Mac, but get the error mentioned above. Any help with that?

**Comment by tanlongzhi** [comment]
You can run k8 on Mac with `k8-Darwin`.

**Comment by tarak77** [comment]
Cool. I generate an out.3dg file for my diploid cell
[out.3dg.zip](https://github.com/lh3/hickit/files/2362702/out.3dg.zip)

But again while converting to` .cif `using `dip-c color` and `dip-c vis`, I get the error:
```
(py2) Taraks-MacBook-Pro:dip-c tarakshisode$ ./dip-c color -n ./NXT-43/mm10.chr.txt ../hickit/1CDX1-25/out.3dg | ./dip-c vis -c /dev/stdin ../hickit/1CDX1-25/out.3dg > ./1CDX1-25/1CDX1-25.20kb.dip-c.cif
-bash: ./1CDX1-25/1CDX1-25.20kb.dip-c.cif: No such file or directory
Traceback (most recent call last):
  File "./dip-c", line 122, in <module>
    main()
  File "./dip-c", line 73, in main
    return_value = color.color(sys.argv[1:])
  File "/Users/tarakshisode/dip-c/color.py", line 127, in color
    g3d_data = file_to_g3d_data(open(args[0], "rb"))
  File "/Users/tarakshisode/dip-c/classes.py", line 1419, in file_to_g3d_data
    g3d_data.add_g3d_particle(string_to_g3d_particle(g3d_file_line.strip()))
  File "/Users/tarakshisode/dip-c/classes.py", line 1214, in string_to_g3d_particle
    hom_name, ref_locus, x, y, z = g3d_particle_string.split("\t")
ValueError: need more than 1 value to unpack
```

**Comment by tanlongzhi** [comment]
Wonderful!

First of all, the sex of this cell (female) was not correctly recognized by the pipeline. According to my understanding, this repo determines sex by the presence of chromosome Y. Therefore, you will need to replace the original part in the pipeline `./k8 hickit.js chronly -` with `./k8 hickit.js chronly -y -` to remove chromosome Y.

Regarding viewing in Pymol, similar to your [haploid thread](https://github.com/lh3/hickit/issues/3#issuecomment-419248913), the header and chromosome naming (this repo uses `chr1a` and `chr1b`, while dip-c uses `chr1(pat)` and `chr1(mat)`) must be modified before running dip-c. In the diploid case, this will be:
```sh
# female
grep -v "^#" hickit.3dg | sed 's/a/(pat)/g; s/b/(mat)/g' > dip-c.3dg

# male
grep -v "^#" hickit.3dg | sed 's/a/(pat)/g; s/b/(mat)/g; s/X/X(mat)/g; s/Y/Y(pat)/g' > dip-c.3dg
```

Your file looks like this in PyMol (sex was set to male):
<img width="889" alt="2018-09-07 9 02 08" src="https://user-images.githubusercontent.com/2436999/45248621-6368d300-b2e1-11e8-8921-163f55262346.png">

The PyMol script to style the above screenshot was:
```python
as sticks, all
set_bond stick_radius, 0.2, all
spectrum b, rainbow, all, 1, 20
```

Apparently something is wrong with the `.3dg` file because the 3D particles are highly non-contiguous. For example, the first few lines are:
```
chr1a	0	-2.136324	0.655881	-0.327950
chr1a	3000000	-2.266238	-0.301880	0.272601
chr1a	3260000	-2.225729	-0.355070	0.189370
chr1a	3280000	-1.953178	-0.709451	0.167003
chr1a	3460000	-1.849689	-0.687710	0.285149
chr1a	3700000	-2.032488	-0.573000	0.798407
chr1a	3720000	-1.997815	-0.563639	0.627355
chr1a	3740000	-1.760033	-0.400391	0.506453
chr1a	3820000	-1.589145	-0.206919	0.066170
chr1a	3900000	-1.728089	-0.160564	-0.031998
```
After you fix the sex, can you attach the `.pairs` file and the `.3dg` file, so we can figure out what went wrong?

**Comment by tarak77** [comment]
Okay, but I have some quick questions:
1. When you output `.3dg` files separately for male and female, I am a bit confused with the male and female `.3dg` files even though each one of them have `mat` and `pat` chromosomes?

After following the steps you mentioned, I get a an output from `hickit.3dg` to `female.dip-c.3dg`, but an error when converting from `hickit.3dg` to `male.dip-c.3dg`(Note that my `hickit.3dg` file does not contain Y chromosome, so I think I get the following error):
```
.
.
[M::color] analyzed 40000 particles (86.75%)
[M::color] analyzed 41000 particles (88.92%)
[M::color] analyzed 42000 particles (91.09%)
[M::color] analyzed 43000 particles (93.26%)
[M::color] analyzed 44000 particles (95.43%)
Traceback (most recent call last):
  File "./dip-c", line 122, in <module>
    main()
  File "./dip-c", line 73, in main
    return_value = color.color(sys.argv[1:])
  File "/Users/tarakshisode/dip-c/color.py", line 191, in color
    color = ref_name_colors[g3d_particle.get_ref_name()]
  File "/Users/tarakshisode/dip-c/classes.py", line 1181, in get_ref_name
    return hom_name_to_ref_name_haplotype(self.hom_name)[0]
  File "/Users/tarakshisode/dip-c/classes.py", line 40, in hom_name_to_ref_name_haplotype
    ref_name, haplotype = hom_name.split("(")
ValueError: too many values to unpack
[M::__main__] command: ./dip-c vis -c /dev/stdin ./1CDX1-25/male.dip-c.3dg
[M::__main__] finished in 3.7 sec

```

2. Also if it helps, Note that my `read1.sai` (which is produced with .sam file)  looks empty?

3. The files you requested for:
[contacts.pairs.gz](https://github.com/lh3/hickit/files/2362934/contacts.pairs.gz)
[impute.pairs.gz](https://github.com/lh3/hickit/files/2362935/impute.pairs.gz)
[out.3dg.zip](https://github.com/lh3/hickit/files/2362936/out.3dg.zip)

[female.dip-c.3dg.zip](https://github.com/lh3/hickit/files/2362938/female.dip-c.3dg.zip)
[1CDX1-25.20kb.female.dip-c.cif.zip](https://github.com/lh3/hickit/files/2362939/1CDX1-25.20kb.female.dip-c.cif.zip)

**Comment by tanlongzhi** [comment]
Thank you for the files.

Sorry about the confusion. I didn't mean outputing female and male files *separately*. Instead, one should pick only *one* line of code suitable for the sex of one's cell. In your case, you will run the female line only:
```sh
# female
grep -v "^#" hickit.3dg | sed 's/a/(pat)/g; s/b/(mat)/g' > dip-c.3dg
```

Your correctly sexed `.cif` file looks similarly fragmented as before:
<img width="972" alt="2018-09-07 10 11 04" src="https://user-images.githubusercontent.com/2436999/45249242-2a356080-b2eb-11e8-879a-15626ed91936.png">

I suspect that the problem is related to filtering out contact-poor 3D particles after 3D modeling, because the number of contacts (only ~ 70 k) is too low compared to your chosen resolution (20 kb). While I look into the filtering, you should consider 3D modeling with lower resolution, for example 100 kb or more. This cell's low number of contacts will not accurately determine high-resolution 3D structures.

**Comment by tarak77** [comment]
Oh I see now.
It might be trivial, but how could you determine the sex of the cell?

**Comment by tanlongzhi** [comment]
Good question. We typically determine the sex of a cell by inferring the copy numbers of chromosomes X and Y from read depths, which is highly accurate for Dip-C and some other methods. In the case of Nagano et al. (2017), however, the cell line is well known to be female and widely used to study X chromosome inactivation, a female-only phenomenon.

In general, in diploid 3D modeling, it is crucial to know the copy numbers of all chromosomes and their homologs. Most cell lines are prone to very large copy number variations (CNVs) and losses of heterozygosity (LOHs). See our supplement and the supplement of Nagano et al. (2017) for examples. Large CNV gains and LOHs must be removed, and CNV losses should be ideally dealt with, prior to imputation and 3D modeling.

**Comment by tarak77** [comment]
I see, I will again go over it tomorrow. Thanks you!

**Comment by tanlongzhi** [comment]
Regarding the loss of many 3D particles in your 20-kb structure, according to my understanding (@lh3 please correct me if I got it wrong), this repo handles particles with no contacts by removing/merging them. This is different from the behavior of dip-c and nuc_dynamics, where contact-poor particles are carried along during 3D modeling, and removed later but only in large regions (e.g. very few contacts in its 1-Mb neighborhood) or manually filtered out later by large RMSD between replicates.

In other words, when the resolution is set too high for the number of contacts in the given single cell, this repo starts to discard 3D particles, while dip-c and nuc_dynamics generate floppy 3D structures with large RMSD. Anyways it's a sign to use a lower resolution.

**Comment by tanlongzhi** [comment]
Btw when I [replied](https://github.com/lh3/hickit/issues/5#issuecomment-419560125) to you a while ago, I copied a typo from the README (`-p` should not be used for `bwa mem` in this case). Both my reply and the README are now corrected (#8). I know you had it correct originally, and hope my reply didn't mislead you.

**Comment by tarak77** [comment]
Got it, thanks for the explanation.
